### DISCUSSION TITLE
Unable to disable local inbox

### DISCUSSION THREAD

**Discussion by Daerendor** [docker-mailserver/docker-mailserver#3472] [Q&A] [answered]
Let me describe my env first.
I have specific MX records on domain.tld that will redirect the emails for domain.tld to a third party service.
For any@domain.tld --> third party inbox
I also have some selfhosted apps, and I'd like to send emails from them.
From noreply@domain.tld to anyone

Some of these emails (especially system-related ones) are sent for me as well. This means that they end up in my DMS inbox (since they are treated as local by DMS). However, I would like to let them be sent just as any other email, since my MX records will make them go to my thid-party inbox.
From noreply@domain.tld To me@domain.tld --> MX --> third party inbox

To achieve this goal, I tried setting $mydestination to localhost, I also tried to leave it empty, but it did not work.
Tl;dr, I am trying to "disable" local inbox.
Any clues?

Note: I am using local auth.

**Comment by polarathene** [comment]
`mydestination` is for Postfix to store in `/var/mail` instead of sending to Dovecot over LMTP to store in a subdir of `/var/mail` which Dovecot manages. The `mydestination` domain is used for system accounts (amavis/clamav/root/etc) and Postfix itself to manage internal mail specific to DMS, not accounts you add to DMS.

A common scenario is `amavis@mail.example.com` (_DMS hostname, no `setup email add`  used_), while the same `example.com` is used explicitly for user mail accounts like `hello@example.com`.

Some users misunderstand this and think configuring a bare domain with the config `hostname: example.com` is a better choice as they believed they were setting up `hello@mail.example.com`, but DMS hostname and email accounts you setup with `setup email add` are unrelated, you cannot overlap them though... so our FAQ entry on bare domains advises modifying `mydestination`, which can cause delivery issues with those internal system/postfix accounts.

---

As for local inbox, it's not clear what you're referring to here. If email is submitted to DMS (Postfix) to send mail it should be via ports 587 or 465, both requiring auth to send mail outbound (to a third-party mail server).

You have your sender (`noreply@example.com`) and your recipient (`me@example.com`). In this case `me@example.com` may be recognized as a local account managed by Dovecot and mail gets delivered there. If you have created an explicit `me@example.com` account that will be the case, otherwise if you have a wildcard virtual alias `@example.com` to redirect to a third-party mail service, then mail should be sent to that, unless a `me@example.com` virtual alias was defined earlier and directed it somewhere else. Virtual aliases (`setup alias add`) have priority over local Dovecot accounts (`setup email add`).

- If what you need is a local `me@example.com` to store your mail in Dovecot sometimes, but send to a third-party when the sender address is something specific (`noreply@example.com`) you could configure Postfix for this but DMS has no official support for such.
- If you have no need for any local storage of mail and always use a third-party service, then `setup alias add` is what you want, not `setup email add`, and a wildcard to modify the recipient domain / address should do the trick (`@example.com` to  `me@gmail.com`).

---

So it sounds like your main issue is you created a `@example.com` local account with `setup email add` where that `example.com` domain is not actually managed by DMS, that is a misconfiguration as you're telling DMS that it manages mail for that domain (should have an MX record pointing to DMS).

If spoof protection is not enabled in DMS ENV, then you can have any `setup email add` account you make (_doesn't have to be for `example.com`_) can send mail as any address (_once authenticated over port 465 or 587_).
- `setup email add admin@example.test`
- Login as `admin@example.test` to DMS at `mail.example.com`
- Send mail as `noreply@example.com`, or even as `john.doe@gmail.com` (_which will fail if `gmail.com` DNS doesn't allow `mail.example.com` to send_)
- Mail is submitted to DMS account `admin@example.test`, and then `mail.example.com` connects to the recipient address (`me@gmail.com`) mail server (MX record) with the mail sender you chose (`noreply@example.com`).

That said, there may be a Postfix setting that would force the MX record lookup, even when there is a local mailbox domain match like `example.com` that Postfix thought it was the final delivery location too. I believe you're trying to configure DMS as a "null mailer" (sender-only service).

**Reply by polarathene** [reply]
Response I wrote a few weeks ago for a potentially similar setup:
https://github.com/docker-mailserver/docker-mailserver/issues/3437#issuecomment-1646350088

If you just want a web UI, you could also have that pull mail from DMS via IMAP/POP3 and submit mail through DMS while composing in the web client. Local app like ThunderBird can work pretty well though too.

**Reply by Daerendor** [reply]
Thank you very much for your reply, lot of useful information here.

Starting from the bottom,
- Yes, since I already have MX records for `example.com` pointing to my third party mailing service, checking the "external" MX even for local accounts is something I'm looking for. And yeah, I'm using DMS as a send-only service, for other docker apps.
- My third party service has a catch-all thing configured, so I really don't need any local inbox.
- Yeah, I've read the docs and other GH issues, the ones that end by someone saying "Add user-patches to change mydestination, but you shouldn't do it". I guess I'm going to try and avoid it.


I tried setting up the recommended way, i.e. not a bare domain.
Something is still unclear to me tho, about the local inbox delivery .


```bash
echo "smth" | mail -s "sub" trash@domain.tld -r noreply@domain.tld
```
```log
Aug 12 01:47:39 mail postfix/pickup[637]: C14D26395E: uid=0 from=<noreply@domain.tld>
Aug 12 01:47:39 mail postfix/cleanup[650]: C14D26395E: message-id=<20230811234739.C14D26395E@mail.domain.tld>
Aug 12 01:47:39 mail opendkim[524]: C14D26395E: DKIM-Signature field added (s=mail, d=domain.tld)
Aug 12 01:47:39 mail postfix/qmgr[638]: C14D26395E: from=<noreply@domain.tld>, size=450, nrcpt=1 (queue active)
Aug 12 01:47:39 mail dovecot: lmtp(644): Connect from local
Aug 12 01:47:39 mail dovecot: auth: passwd-file(trash@domain.tld): unknown user
Aug 12 01:47:39 mail postfix/lmtp[640]: C14D26395E: to=<trash@domain.tld>, relay=mail.domain.tld[/var/run/dovecot/lmtp], delay=0.1, delays=0.09/0/0/0, dsn=5.1.1, status=bounced (host mail.domain.tld[/var/run/dovecot/lmtp] said: 550 5.1.1 <trash@domain.tld> User doesn't exist: trash@domain.tld (in reply to RCPT TO command))
Aug 12 01:47:39 mail dovecot: lmtp(644): Disconnect from local: Logged out (state=READY)
Aug 12 01:47:39 mail postfix/cleanup[648]: D4D5563960: message-id=<20230811234739.D4D5563960@mail.domain.tld>
Aug 12 01:47:39 mail postfix/bounce[645]: C14D26395E: sender non-delivery notification: D4D5563960
Aug 12 01:47:39 mail postfix/qmgr[638]: D4D5563960: from=<>, size=3201, nrcpt=1 (queue active)
Aug 12 01:47:39 mail postfix/qmgr[638]: C14D26395E: removed
Aug 12 01:47:39 mail dovecot: lmtp(642): Connect from local
Aug 12 01:47:39 mail dovecot: lmtp(noreply@domain.tld)<642><OEJqNJvI1mSCAgAAvaCfBw>: sieve: msgid=<20230811234739.D4D5563960@mail.domain.tld>: stored mail into mailbox 'INBOX'
Aug 12 01:47:39 mail postfix/lmtp[641]: D4D5563960: to=<noreply@domain.tld>, relay=mail.domain.tld[/var/run/dovecot/lmtp], delay=0.02, delays=0.01/0/0/0.01, dsn=2.0.0, status=sent (250 2.0.0 <noreply@domain.tld> OEJqNJvI1mSCAgAAvaCfBw Saved)
Aug 12 01:47:39 mail postfix/qmgr[638]: D4D5563960: removed
Aug 12 01:47:39 mail dovecot: lmtp(642): Disconnect from local: Logged out (state=READY)
```
```env
OVERRIDE_HOSTNAME=
LOG_LEVEL=debug
SUPERVISOR_LOGLEVEL=debug
ONE_DIR=1
ACCOUNT_PROVISIONER=FILE
TLS_LEVEL=modern
SPOOF_PROTECTION=0
ENABLE_SRS=0
ENABLE_OPENDKIM=1
ENABLE_OPENDMARC=1
ENABLE_POLICYD_SPF=1
ENABLE_[OTHER SERVICES]=0
SMTP_ONLY=0
ENABLE_SASLAUTHD=1
SASLAUTHD_MECHANISMS=shadow
SRS_SENDER_CLASSES=envelope_sender
```
So, based on your recommendations, I configured DMS to use mail.domain.tld, and then I added a @domain.tld user.
Since @domain.tld is not local to mail.domain.tld, it shouldn't be flagged as local delivery, right? So why is this happening? Is this related to the "force MX check" you were talking about?

**Reply by polarathene** [reply]
## Breaking down the log

> Since `@domain.tld` is not local to `mail.domain.tld`, it shouldn't be flagged as local delivery, right?

```
Aug 12 01:47:39 mail postfix/pickup[637]: C14D26395E: uid=0 from=<noreply@domain.tld>
Aug 12 01:47:39 mail postfix/cleanup[650]: C14D26395E: message-id=<20230811234739.C14D26395E@mail.domain.tld>
Aug 12 01:47:39 mail opendkim[524]: C14D26395E: DKIM-Signature field added (s=mail, d=domain.tld)
Aug 12 01:47:39 mail postfix/qmgr[638]: C14D26395E: from=<noreply@domain.tld>, size=450, nrcpt=1 (queue active)
```

- The mail has arrived to DMS (Postfix) from the sender `noreply@domain.tld`.
- It is assigned a `message-id` and signed with DKIM key.
- Message (email) is added to the queue to be handled.

I don't think it should be signed by DKIM unless it's aware it's sending/relaying the mail, not just receiving it 🤔 

```
Aug 12 01:47:39 mail dovecot: lmtp(644): Connect from local
Aug 12 01:47:39 mail dovecot: auth: passwd-file(trash@domain.tld): unknown user
Aug 12 01:47:39 mail postfix/lmtp[640]: C14D26395E: to=<trash@domain.tld>, relay=mail.domain.tld[/var/run/dovecot/lmtp], delay=0.1, delays=0.09/0/0/0, dsn=5.1.1, status=bounced (host mail.domain.tld[/var/run/dovecot/lmtp] said: 550 5.1.1 <trash@domain.tld> User doesn't exist: trash@domain.tld (in reply to RCPT TO command))
```

- Now Postfix has connected to the virtual transport over LMTP, which is to Dovecot for local delivery. At least when it's for the domain it believes it owns fully for delivery (`@domain.tld`).
- I think at this stage Postfix is checking with Dovecot if an actual mail account sends before continuing. In this case it gets told this user is not known, there is no inbox to delivery to.
- Since Postfix thinks that Dovecot is the final destination, it decides something is not right and chooses to send a bounce email back to the sender `noreply@domain.tld`.

```
Aug 12 01:47:39 mail dovecot: lmtp(642): Connect from local
Aug 12 01:47:39 mail dovecot: lmtp(noreply@domain.tld)<642><OEJqNJvI1mSCAgAAvaCfBw>: sieve: msgid=<20230811234739.D4D5563960@mail.domain.tld>: stored mail into mailbox 'INBOX'
Aug 12 01:47:39 mail postfix/lmtp[641]: D4D5563960: to=<noreply@domain.tld>, relay=mail.domain.tld[/var/run/dovecot/lmtp], delay=0.02, delays=0.01/0/0/0.01, dsn=2.0.0, status=sent (250 2.0.0 <noreply@domain.tld> OEJqNJvI1mSCAgAAvaCfBw Saved)
```

Now the same process happens, but in this case you have created a `noreply@domain.tld` account (`setup email add`?), so it is accepted and delivered to local storage managed by Dovecot, not your third-party service you would have wanted.

---

## Avoiding local delivery (send-only)

> Is this related to the "force MX check" you were talking about?

Actually, since you don't need Dovecot for any local storage right? You can just disable Dovecot, it is not doing anything for you from what I understand?

The earlier mentioned system accounts that mail to `@mail.domain.tld` by default will still be handled by Postfix storing in `/var/mail`, unless you setup proper alias config to route to your `@domain.tld` address, which should then get delivered to the third-party mail service after solving the below concern.

### `SMTP_ONLY=1` (when you don't need Dovecot / mail storage)

We have [`SMTP_ONLY=1`](https://docker-mailserver.github.io/docker-mailserver/v12.1/config/environment/#smtp_only) for this, it was implemented before I became a maintainer of the project.

That configures Postfix to disable SASL auth unless you're using `ENABLE_SASLAUTHD=1`:

https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/startup/setup.d/postfix.sh#L25-L40

And disables setting up Dovecot and the local user database (`ACCOUNT_PROVISIONER=FILE`):

https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/start-mailserver.sh#L45-L56

https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/startup/setup.d/dovecot.sh#L136-L138

https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/check-for-changes.sh#L104-L105

https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/helpers/accounts.sh#L9-L18

https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/scripts/check-for-changes.sh#L57

---

### Authenticating with `SMTP_ONLY=1` (Probably with `ENABLE_SASLAUTHD=1`)

Easiest may be with LDAP, but there are a few other options.

However, the contributor that introduced this feature I believe introduced the SASL provider alternative (instead of from Dovecot), SASLAuthD. So there is extra setup required, but the original Wiki never had this information contributed (_our current docs didn't exist at the time, so wasn't part of the PR process_).

We have a [FAQ entry for `SMTP_ONLY=1`](https://docker-mailserver.github.io/docker-mailserver/v12.1/faq/#how-can-i-authenticate-users-with-smtp_only1) that [references an issue discussing some approaches](https://github.com/docker-mailserver/docker-mailserver/issues/1247) to add users to authenticate with for mail submission. For context, in 2019 Cyrus (_**EDIT:** might have been Courier_) was used instead of Dovecot IIRC, and while the user starts with `ENABLE_SASLAUTHD=1`, it looks like they use a different config approach? (_**EDIT:** maybe not, seems the [Postfix docs for Cyrus SASL is related to saslauthd](http://www.postfix.org/SASL_README.html#saslauthd)_) And earlier this year another user shared their `user-patches.sh` snippet (_looks like it has a typo, probably should be `pwcheck_method`, not `wcheck_method`_).

There's also a brief mention in a user contributed guide for a [relay server setup using `SMTP_ONLY=1`](https://docker-mailserver.github.io/docker-mailserver/v12.1/examples/use-cases/forward-only-mailserver-with-ldap-authentication/#building-a-forward-only-mail-server), this highlights a few other services that aren't relevant for a server just passing on mail (_and presumably applicable when only sending too_), so could be disabled. `SMTP_ONLY=1` could probably be updated to also disable those services 🤷‍♂️ 

Our [test coverage for the `SMTP_ONLY=1` feature](https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/test/tests/parallel/set3/mta/smtponly.bats) doesn't look too great though. For the `ENABLE_SASLAUTHD=1` feature, the contributor only introduced it for connecting to their LDAP provider, and for some reason they also have a [`rimap` provider test](https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/test/tests/serial/mail_with_imap.bats) (_not sure what that is, but since `SMTP_ONLY=1` is not used, it may be an alternative to use Dovecot SASL_), here's the related config file if you choose to look into `ENABLE_SASLAUTHD=1`:

https://github.com/docker-mailserver/docker-mailserver/blob/f28fce9cc432f1f447bd963d9e54e44bcf2c27dd/target/supervisor/conf.d/saslauth.conf#L28-L35

**EDIT:** Yeah, according to the note in this LDAP user contributed guide, the [`rimap` saslauthd provider seems to use Dovecot](https://docker-mailserver.github.io/docker-mailserver/v12.1/config/advanced/auth-ldap/#saslauthd_ldap_filter). It seems to be a generic IMAP provider solution instead of the Postfix support for Dovecot directly as a SASL provider.

These are features that no active maintainer is that familiar with, nor has too much time to delve into unfortunately. So their maintenance is a bit lackluster.

---

### Alternative: Disabling the Postfix config related to Dovecot delivery, but keeping the Dovecot SASL auth integration enabled

Without `SMTP_ONLY=1`, it might be possible to override Postfix config with [`postfix-main.cf` and `postfix-master.cf`](https://docker-mailserver.github.io/docker-mailserver/v12.1/config/advanced/override-defaults/postfix/), to still leverage the account management with Dovecot, but opt-out of Postfix using it for delivery, only authentication to send mail.

Off the top of my head I don't know the exact settings, I think it's mostly the `virtual_` prefix for delivery to Dovecot. Our scripts also probably parse the accounts and aliases config to extract domains which may be relevant. So we could improve on this, but presently it may not be any easier than the other alternative auth solutions.

**Reply by polarathene** [reply]
**TL;DR:** You probably want `SMTP_ONLY=1`, it'll opt-out of support to receive mail.

This may break authentication support needed to send mail outbound (to your service). Looks like that can be resolved by enabling LDAP (_problematic for some users, no real maintainer support in DMS_), or `ENABLE_SASLAUTHD=1` with some extra config that DMS has no official docs / support for at present.

Raised a TODO task issue, hopefully someone may find time to contribute improvements to supporting this usage better with DMS: https://github.com/docker-mailserver/docker-mailserver/issues/3477

**Reply by Daerendor** [reply, marked as answer]
I managed to get it working, thanks to [this issue](https://github.com/docker-mailserver/docker-mailserver/issues/1247) you told me to take a look at.

`mail.env`
```env
SMTP_ONLY=1
SMTP_USERNAME=user
SMTP_DOMAIN=domain.tld
SMTP_PASSWORD=password
ENABLE_SASLAUTHD=1
SASLAUTHD_MECHANISMS=
```

`postfix-main.cf`
```bash
smtpd_sasl_auth_enable = yes
smtpd_sasl_path = smtpd
smtpd_sasl_type = cyrus
```

`user-patches.sh`
```bash
#!/bin/bash

mkdir -p /etc/postfix/sasl

echo "pwcheck_method: auxprop
auxprop_plugin: sasldb
mech_list: PLAIN" > /etc/postfix/sasl/smtpd.conf

echo $SMTP_PASSWORD | saslpasswd2 -c -u $SMTP_DOMAIN $SMTP_USERNAME
```

I believe $SMTP_DOMAIN is useless, but I am not a pro at this so I'll leave it like this for now.

Adding a new 'local' mechanism to SASLAUTD_MECHANISMS, doing exactly this could be an idea. I'll leave it to you to update the todo you've created.

Thanks!
